jobrunner/schema.py

```diff
@@ -0,0 +1,43 @@
+from dataclasses import dataclass
+from enum import Enum
+
+from jobrunner.job_executor import ExecutorState
+
+
+class TaskType(Enum):
+    RUNJOB = "runjob"
+    # TODO: delete job
+
+
+class TaskStage(Enum):
+    """Valid stage of any tasks
+
+    A big merge of multiple types of TaskType's stages
+    """
+
+    # RUNJOB stages
+    PREPARING = ExecutorState.PREPARING.value
+    PREPARED = ExecutorState.PREPARED.value
+    EXECUTING = ExecutorState.EXECUTING.value
+    EXECUTED = ExecutorState.EXECUTED.value
+    FINALIZING = ExecutorState.FINALIZING.value
+    # Final stages
+    FINALIZED = ExecutorState.FINALIZED.value
+    ERROR = ExecutorState.ERROR.value
+
+
+@dataclass(frozen=True)
+class AgentTask:
+    """Task API task data
+
+    This is basically all the information about the task supplied by the
+    controller, and is effectively immutable.
+
+    The agent's view of a task's current state is computed on demand from
+    disk/docker state
+    """
+
+    id: str  # noqa: A003
+    type: TaskType  # noqa: A003
+    definition: dict
+    created_at: int = None
```

Comment:
should we also store `State` & `StatusCode` here?

Reply:
I don't think so, as the State will never change from the agent's POV (it's always RUNNING), and StatusCode is the controller's responsibility, agent should never worry about it.

Reply:
The presences of agent_results is the primary indicated that a task has finished:

https://github.com/opensafely-core/job-runner/pull/821/files#diff-c71376d8e6de22a033c81b0636cc27f102c41485bc40d738fcaca92176d61a01R419

Additionally, I do not think is not going to be possible in step 1 to make that TaskStage opaque to the controller, as it will need to set the relevant StatusCode in the controller DB in order to maintain the current behaviour.

Once we are free to change things, there may be a way to make them opaque, perhaps.